File: scripts/service_manager.py
#!/usr/bin/env python3
"""
Chaos World Service Manager
A Python-based service management tool for Windows services using NSSM
"""

import subprocess
import sys
import time
import requests
import json
from pathlib import Path

class ServiceManager:
    def __init__(self):
        self.services = {
            'api-gateway': 'ChaosWorld-API-Gateway',
            'chaos-backend': 'ChaosWorld-Backend'
        }
        self.endpoints = {
            'api-gateway': 'http://localhost:8080',
            'chaos-backend': 'http://localhost:8081'
        }
    
    def run_command(self, command, check=True):
        """Run a command and return the result"""
        try:
            result = subprocess.run(command, shell=True, capture_output=True, text=True, check=check)
            return result.returncode == 0, result.stdout, result.stderr
        except subprocess.CalledProcessError as e:
            return False, e.stdout, e.stderr
    
    def check_service_status(self, service_name):
        """Check if a service is running"""
        success, stdout, stderr = self.run_command(f'sc query "{service_name}"', check=False)
        if success:
            return 'RUNNING' in stdout
        return False
    
    def start_service(self, service_name):
        """Start a service"""
        print(f"🚀 Starting {service_name}...")
        success, stdout, stderr = self.run_command(f'nssm start "{service_name}"', check=False)
        if success:
            print(f"✅ {service_name} started successfully")
            return True
        else:
            print(f"❌ Failed to start {service_name}: {stderr}")
            return False
    
    def stop_service(self, service_name):
        """Stop a service"""
        print(f"🛑 Stopping {service_name}...")
        success, stdout, stderr = self.run_command(f'nssm stop "{service_name}"', check=False)
        if success:
            print(f"✅ {service_name} stopped successfully")
            return True
        else:
            print(f"❌ Failed to stop {service_name}: {stderr}")
            return False
    
    def restart_service(self, service_name):
        """Restart a service"""
        print(f"🔄 Restarting {service_name}...")
        self.stop_service(service_name)
        time.sleep(2)
        return self.start_service(service_name)
    
    def test_endpoint(self, service_name, endpoint):
        """Test if an endpoint is responding"""
        try:
            response = requests.get(f"{endpoint}/health", timeout=5)
            if response.status_code == 200:
                print(f"✅ {service_name} is responding on {endpoint}")
                return True
            else:
                print(f"❌ {service_name} returned status {response.status_code}")
                return False
        except requests.exceptions.RequestException as e:
            print(f"❌ {service_name} is not responding: {e}")
            return False
    
    def check_port(self, port):
        """Check if a port is in use"""
        success, stdout, stderr = self.run_command(f'netstat -an | findstr ":{port}"', check=False)
        return success
    
    def status(self):
        """Show comprehensive status of all services"""
        print("=" * 60)
        print("🔍 CHAOS WORLD SERVICES STATUS")
        print("=" * 60)
        
        for service_key, service_name in self.services.items():
            print(f"\n📋 {service_key.upper()}:")
            is_running = self.check_service_status(service_name)
            port = 8080 if service_key == 'api-gateway' else 8081
            port_in_use = self.check_port(port)
            endpoint_responding = self.test_endpoint(service_key, self.endpoints[service_key])
            
            status_icon = "✅" if is_running else "❌"
            port_icon = "✅" if port_in_use else "❌"
            endpoint_icon = "✅" if endpoint_responding else "❌"
            
            print(f"   Service: {status_icon} {'RUNNING' if is_running else 'STOPPED'}")
            print(f"   Port {port}: {port_icon} {'IN USE' if port_in_use else 'NOT IN USE'}")
            print(f"   Endpoint: {endpoint_icon} {'RESPONDING' if endpoint_responding else 'NOT RESPONDING'}")
    
    def start_all(self):
        """Start all services"""
        print("🚀 Starting all Chaos World services...")
        success = True
        for service_key, service_name in self.services.items():
            if not self.start_service(service_name):
                success = False
        return success
    
    def stop_all(self):
        """Stop all services"""
        print("🛑 Stopping all Chaos World services...")
        success = True
        for service_key, service_name in self.services.items():
            if not self.stop_service(service_name):
                success = False
        return success
    
    def restart_all(self):
        """Restart all services"""
        print("🔄 Restarting all Chaos World services...")
        self.stop_all()
        time.sleep(3)
        return self.start_all()

def main():
    manager = ServiceManager()
    
    if len(sys.argv) < 2:
        print("Usage: python service_manager.py <command>")
        print("Commands:")
        print("  status     - Show service status")
        print("  start      - Start all services")
        print("  stop       - Stop all services")
        print("  restart    - Restart all services")
        print("  start-api  - Start API Gateway only")
        print("  start-backend - Start Chaos Backend only")
        print("  stop-api   - Stop API Gateway only")
        print("  stop-backend - Stop Chaos Backend only")
        print("  restart-api - Restart API Gateway only")
        print("  restart-backend - Restart Chaos Backend only")
        sys.exit(1)
    
    command = sys.argv[1].lower()
    
    if command == 'status':
        manager.status()
    elif command == 'start':
        manager.start_all()
    elif command == 'stop':
        manager.stop_all()
    elif command == 'restart':
        manager.restart_all()
    elif command == 'start-api':
        manager.start_service(manager.services['api-gateway'])
    elif command == 'start-backend':
        manager.start_service(manager.services['chaos-backend'])
    elif command == 'stop-api':
        manager.stop_service(manager.services['api-gateway'])
    elif command == 'stop-backend':
        manager.stop_service(manager.services['chaos-backend'])
    elif command == 'restart-api':
        manager.restart_service(manager.services['api-gateway'])
    elif command == 'restart-backend':
        manager.restart_service(manager.services['chaos-backend'])
    else:
        print(f"Unknown command: {command}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: scripts/service_manager.js
#!/usr/bin/env node
/**
 * Chaos World Service Manager
 * A Node.js-based service management tool for Windows services using NSSM
 */

const { exec, spawn } = require('child_process');
const http = require('http');
const util = require('util');

const execAsync = util.promisify(exec);

class ServiceManager {
    constructor() {
        this.services = {
            'api-gateway': 'ChaosWorld-API-Gateway',
            'chaos-backend': 'ChaosWorld-Backend'
        };
        this.endpoints = {
            'api-gateway': 'http://localhost:8080',
            'chaos-backend': 'http://localhost:8081'
        };
    }

    async runCommand(command, check = true) {
        try {
            const { stdout, stderr } = await execAsync(command);
            return { success: true, stdout, stderr };
        } catch (error) {
            if (check) {
                return { success: false, stdout: error.stdout, stderr: error.stderr };
            }
            return { success: true, stdout: error.stdout, stderr: error.stderr };
        }
    }

    async checkServiceStatus(serviceName) {
        const result = await this.runCommand(`sc query "${serviceName}"`, false);
        if (result.success) {
            return result.stdout.includes('RUNNING');
        }
        return false;
    }

    async startService(serviceName) {
        console.log(`🚀 Starting ${serviceName}...`);
        const result = await this.runCommand(`nssm start "${serviceName}"`, false);
        if (result.success) {
            console.log(`✅ ${serviceName} started successfully`);
            return true;
        } else {
            console.log(`❌ Failed to start ${serviceName}: ${result.stderr}`);
            return false;
        }
    }

    async stopService(serviceName) {
        console.log(`🛑 Stopping ${serviceName}...`);
        const result = await this.runCommand(`nssm stop "${serviceName}"`, false);
        if (result.success) {
            console.log(`✅ ${serviceName} stopped successfully`);
            return true;
        } else {
            console.log(`❌ Failed to stop ${serviceName}: ${result.stderr}`);
            return false;
        }
    }

    async restartService(serviceName) {
        console.log(`🔄 Restarting ${serviceName}...`);
        await this.stopService(serviceName);
        await new Promise(resolve => setTimeout(resolve, 2000));
        return await this.startService(serviceName);
    }

    async testEndpoint(serviceName, endpoint) {
        return new Promise((resolve) => {
            const req = http.get(`${endpoint}/health`, (res) => {
                if (res.statusCode === 200) {
                    console.log(`✅ ${serviceName} is responding on ${endpoint}`);
                    resolve(true);
                } else {
                    console.log(`❌ ${serviceName} returned status ${res.statusCode}`);
                    resolve(false);
                }
            });
            
            req.on('error', (err) => {
                console.log(`❌ ${serviceName} is not responding: ${err.message}`);
                resolve(false);
            });
            
            req.setTimeout(5000, () => {
                console.log(`❌ ${serviceName} request timed out`);
                req.destroy();
                resolve(false);
            });
        });
    }

    async checkPort(port) {
        const result = await this.runCommand(`netstat -an | findstr ":${port}"`, false);
        return result.success;
    }

    async status() {
        console.log('='.repeat(60));
        console.log('🔍 CHAOS WORLD SERVICES STATUS');
        console.log('='.repeat(60));
        
        for (const [serviceKey, serviceName] of Object.entries(this.services)) {
            console.log(`\n📋 ${serviceKey.toUpperCase()}:`);
            const isRunning = await this.checkServiceStatus(serviceName);
            const port = serviceKey === 'api-gateway' ? 8080 : 8081;
            const portInUse = await this.checkPort(port);
            const endpointResponding = await this.testEndpoint(serviceKey, this.endpoints[serviceKey]);
            
            const statusIcon = isRunning ? '✅' : '❌';
            const portIcon = portInUse ? '✅' : '❌';
            const endpointIcon = endpointResponding ? '✅' : '❌';
            
            console.log(`   Service: ${statusIcon} ${isRunning ? 'RUNNING' : 'STOPPED'}`);
            console.log(`   Port ${port}: ${portIcon} ${portInUse ? 'IN USE' : 'NOT IN USE'}`);
            console.log(`   Endpoint: ${endpointIcon} ${endpointResponding ? 'RESPONDING' : 'NOT RESPONDING'}`);
        }
    }

    async startAll() {
        console.log('🚀 Starting all Chaos World services...');
        let success = true;
        for (const [serviceKey, serviceName] of Object.entries(this.services)) {
            if (!(await this.startService(serviceName))) {
                success = false;
            }
        }
        return success;
    }

    async stopAll() {
        console.log('🛑 Stopping all Chaos World services...');
        let success = true;
        for (const [serviceKey, serviceName] of Object.entries(this.services)) {
            if (!(await this.stopService(serviceName))) {
                success = false;
            }
        }
        return success;
    }

    async restartAll() {
        console.log('🔄 Restarting all Chaos World services...');
        await this.stopAll();
        await new Promise(resolve => setTimeout(resolve, 3000));
        return await this.startAll();
    }
}

async function main() {
    const manager = new ServiceManager();
    const command = process.argv[2];

    if (!command) {
        console.log('Usage: node service_manager.js <command>');
        console.log('Commands:');
        console.log('  status     - Show service status');
        console.log('  start      - Start all services');
        console.log('  stop       - Stop all services');
        console.log('  restart    - Restart all services');
        console.log('  start-api  - Start API Gateway only');
        console.log('  start-backend - Start Chaos Backend only');
        console.log('  stop-api   - Stop API Gateway only');
        console.log('  stop-backend - Stop Chaos Backend only');
        console.log('  restart-api - Restart API Gateway only');
        console.log('  restart-backend - Restart Chaos Backend only');
        process.exit(1);
    }

    switch (command) {
        case 'status':
            await manager.status();
            break;
        case 'start':
            await manager.startAll();
            break;
        case 'stop':
            await manager.stopAll();
            break;
        case 'restart':
            await manager.restartAll();
            break;
        case 'start-api':
            await manager.startService(manager.services['api-gateway']);
            break;
        case 'start-backend':
            await manager.startService(manager.services['chaos-backend']);
            break;
        case 'stop-api':
            await manager.stopService(manager.services['api-gateway']);
            break;
        case 'stop-backend':
            await manager.stopService(manager.services['chaos-backend']);
            break;
        case 'restart-api':
            await manager.restartService(manager.services['api-gateway']);
            break;
        case 'restart-backend':
            await manager.restartService(manager.services['chaos-backend']);
            break;
        default:
            console.log(`Unknown command: ${command}`);
            process.exit(1);
    }
}

if (require.main === module) {
    main().catch(console.error);
}
